### PR TITLE
[MRG] reorder argument for matrix plotting function

### PR DESCRIPTION
**Pull request by mjboos**
Hey y'all,

this is the pull request for issue #1543 
I added a reorder argument to the plot_matrix function. Default (for reorder=True) is average linkage.
A quick comparison on the probabilistic atlas example, seems to give the most sensible ("blockiest") ordering for average and complete. 
different_linkages

Should this be tested on more/different data?

### CODE REVIEW

**Comment by bthirion** [comment]
Thta's fine. actually the average and complete seem just to be the same thing flipped.

File: nilearn/plotting/matrix_plotting.py [hunk 1]
```diff
@@ -62,11 +62,40 @@ def plot_matrix(mat, title=None, labels=None, figure=None, axes=None,
...
 
         Returns Matplotlib AxesImage instance
     """
+    if reorder is not False:
+        if labels is None or labels is False:
+            raise ValueError("Labels are needed to show the reordering.")
+        try:
+            from scipy.cluster.hierarchy import (linkage, optimal_leaf_ordering,
+                                                 leaves_list)
+        except ImportError:
```
Comment (bthirion) [inline nilearn/plotting/matrix_plotting.py]: This has to be enforced in the dependences of the apckage, not a the level of individual functions.

File: nilearn/plotting/tests/test_matrix_plotting.py [hunk 2]
```diff
@@ -12,13 +12,14 @@
...
     plt.close()
     # test if plotting lower triangle works
     ax = plot_matrix(mat, labels=labels, tri='lower')
     # test if it returns an AxesImage
     ax.axes.set_title('Title')
     plt.close()
-
-
+    # test if reordering with default linkage works
+    ax = plot_matrix(mat, labels=labels, reorder=True)
```
Comment (bthirion) [inline nilearn/plotting/tests/test_matrix_plotting.py]: I think we should have a more specific test checking whether the reordering is sensible.

File: nilearn/plotting/matrix_plotting.py (hunk 1 above)
Comment (GaelVaroquaux) [reply nilearn/plotting/matrix_plotting.py]: I disagree: the error is raised only of the user tries to reorder, which is not on by default.

File: examples/03_connectivity/plot_probabilistic_atlas_extraction.py [hunk 3]
```diff
@@ -62,8 +62,10 @@
 from nilearn import plotting
 # Mask out the major diagonal
 np.fill_diagonal(correlation_matrix, 0)
+# We reorder the correlation matrix to make the block structure
+# of invidiual regions more visible.
 plotting.plot_matrix(correlation_matrix, labels=labels, colorbar=True,
-                     vmax=0.8, vmin=-0.8)
+                     vmax=0.8, vmin=-0.8, reorder=True)
```
Comment (GaelVaroquaux) [inline examples/03_connectivity/plot_probabilistic_atlas_extraction.py]: The MSDL atlas is actually ordered by network, so I am not sure that this is the best example to add reordering :)

Comment (GaelVaroquaux) [reply examples/03_connectivity/plot_probabilistic_atlas_extraction.py]: I would try on the following example:
http://nilearn.github.io/auto_examples/03_connectivity/plot_signal_extraction.html

The ordering of the matrix is a bit ugly.

**Comment by GaelVaroquaux** [comment]
Thanks! The only comment that I have is on which example we should demo this feature. Aside from that, +1 for merge.

**Comment by GaelVaroquaux** [comment]
Actually: one important point: you need to skip the reordering test when the scipy version is too old.

File: examples/03_connectivity/plot_probabilistic_atlas_extraction.py (hunk 3 above)
Comment (mjboos) [reply examples/03_connectivity/plot_probabilistic_atlas_extraction.py]: Okay! Worked well, so I added it to this example.

File: nilearn/plotting/tests/test_matrix_plotting.py (hunk 2 above)
Comment (mjboos) [reply nilearn/plotting/tests/test_matrix_plotting.py]: I built now a matrix with some similarities and check if the block structure is present in the reordered labels of the returned AxesImage. A bit convoluted, but it checks if the clustering worked.

**Comment by bthirion** [comment]
LGTM

**Comment by KamalakerDadi** [comment]
Is it still WIP ?

**Comment by mjboos** [comment]
I think it's finished.

**Comment by mjboos** [comment]
Thanks for the comments @KamalakerDadi (and everyone else) I added the tests and notes.

**Comment by KamalakerDadi** [comment]
I have few documentation nitpicks like adding an entry into doc/whats_new and note added at the argument ```reorder``` does not render well. May you should leave one line space above note and indent the note you have written.

These are just minor comments which I think I can address them on the flow. LGTM.

**Comment by KamalakerDadi** [comment]
Thank you for your work.

**Comment by GaelVaroquaux** [comment]
We should finish this PR and merge it. @KamalakerDadi : do you think that we can merge, and you address the small issues with another PR?

**Comment by KamalakerDadi** [comment]
>do you think that we can merge, and you address the small issues with another PR?

Yes. I can do that.

**Pull request merged**